mlir/include/mlir/Dialect/XeGPU/IR/XeGPUTypes.td

```diff
@@ -0,0 +1,170 @@
+    // It will be distributed accross 16 hardware threads, organized as [2, 8],
+    // and each access 2 continious elements in dim 1.
+    xegpu.tensor_desc<8x16xf16, #xegpu.sg_map<wi_layout = [2, 8], wi_data = [1, 2]>>
+
+    // A TensorDesc with 8x16 f32 elements for a memory region in shared memory space.
+    xegpu.tensor_desc<8x16xf32, #xegpu.tdesc_attr<memory_scope = slm>>
+    ```
+  }];
+
+  let parameters = (ins ArrayRefParameter<"int64_t">: $shape,
+                        "mlir::Type": $elementType,
+                        OptionalParameter<"mlir::Attribute">: $encoding);
+
+  let builders = [
+    TypeBuilderWithInferredContext<(ins
+      "llvm::ArrayRef<int64_t>":$shape, "mlir::Type":$elementType,
+      CArg<"mlir::Attribute", "{}"> : $encoding
+    )>,
+    TypeBuilder<(ins
+      "llvm::ArrayRef<int64_t>": $shape, "mlir::Type": $elementType,
+      "mlir::xegpu::MemoryScopeKind": $memory_scope, "int": $array_length,
+      "bool": $boundary_check, "mlir::xegpu::ScatteredAttr": $scattered,
+      "mlir::xegpu::SubGroupMapAttr": $mapping
+    )>,
+    TypeBuilderWithInferredContext<(ins
+      "llvm::ArrayRef<int64_t>": $shape, "mlir::Type": $elementType,
+      "mlir::xegpu::MemoryScopeKind": $memory_scope, "int": $array_length,
+      "bool": $boundary_check, "mlir::xegpu::ScatteredAttr": $scattered,
+      "mlir::xegpu::SubGroupMapAttr": $mapping
+    )>
+  ];
+
+  let extraClassDeclaration = [{
+    using TensorType::clone;
+    using mlir::ShapedType::Trait<TensorDescType>::getElementTypeBitWidth;
+    using mlir::ShapedType::Trait<TensorDescType>::getRank;
+    using mlir::ShapedType::Trait<TensorDescType>::getNumElements;
+    using mlir::ShapedType::Trait<TensorDescType>::isDynamicDim;
+    using mlir::ShapedType::Trait<TensorDescType>::hasStaticShape;
+    using mlir::ShapedType::Trait<TensorDescType>::getNumDynamicDims;
+    using mlir::ShapedType::Trait<TensorDescType>::getDimSize;
+    using mlir::ShapedType::Trait<TensorDescType>::getDynamicDimIndex;
+
+    TensorDescType clone(::mlir::Type elementType) {
+      return llvm::cast<TensorDescType>(cloneWith(getShape(), elementType));
+    }
+
+    TensorDescAttr getEncodingAsTensorDescAttr() const {
+      return llvm::dyn_cast_if_present<TensorDescAttr>(getEncoding());
+    }
+
+    SubGroupMapAttr getEncodingAsMapAttr() const {
+      return llvm::dyn_cast_if_present<SubGroupMapAttr>(getEncoding());
+    }
+
+    ScatteredAttr getEncodingAsScatteredAttr() const {
+      return llvm::dyn_cast_if_present<ScatteredAttr>(getEncoding());
+    }
+
+    xegpu::MemoryScopeKind getMemoryScope();
+    int getArrayLength();
+    bool getBoundaryCheck();
+    xegpu::ScatteredAttr getScattered();
+    xegpu::SubGroupMapAttr getMapping();
+  }];
+
+  let hasCustomAssemblyFormat = true;
+}
+
+
+def XeGPU_Nbarrier: XeGPUTypeDef<"Nbarrier", "nbarrier", [], "mlir::Type"> {
+  let summary = "!xegpu.nbarrier a custom XeGPU type representing a barrier.";
```

Comment:
Seems like maybe where everything about barriers should be documented, a bit like the tensordesc above.